src/planet-a/SubdivSkeleton.h

```diff
@@ -0,0 +1,392 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <osp/id_registry.h>
+
+#include <Corrade/Containers/ArrayViewStl.h>
+
+#include <optional>
+#include <limits>
+#include <unordered_map>
+#include <vector>
+
+namespace planeta
+{
+
+template<typename T>
+using ArrayView_t = Corrade::Containers::ArrayView<T>;
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief A multitree directed acyclic graph of reusable IDs where new IDs can
+ *        be created from two other parent IDs.
+ */
+template<typename ID_T>
+class SubdivIdTree : private osp::IdRegistry<ID_T>
```

Comment:
What's the benfit of allowing different id types?

E.g. why would someone want size_t and uint8_t in the same application?

Reply:
Probably none. this is an over-generalization

Reply:
Actually, I misunderstood this question. Different enum ID types in general are difficult to mix up and can't be implicitly converted. Making all of them uint32_t wouldn't really cause any problems at this point.